File: services/intelligence/app/integrations/slack.py
"""Slack Incoming Webhook notifier for Tombstone events."""

import logging
import os
from typing import Any

import httpx

logger = logging.getLogger(__name__)


class SlackNotifier:
    """Posts Tombstone notifications to a Slack channel via Incoming Webhooks."""

    def __init__(self, webhook_url: str | None = None) -> None:
        self._webhook_url = webhook_url or os.environ.get("SLACK_WEBHOOK_URL", "")

    async def notify_rollback(
        self,
        flag_key: str,
        environment: str,
        error_rate: float,
        triggered_by: str,
        rollback_url: str,
    ) -> bool:
        """Post a red attachment announcing an auto-rollback event.

        Args:
            flag_key: The flag that was disabled.
            environment: Deployment environment (e.g. "production").
            error_rate: Observed error rate that triggered the rollback (0-100).
            triggered_by: Actor / system that initiated the rollback.
            rollback_url: Deep-link into the Tombstone dashboard for this flag.

        Returns:
            True if Slack accepted the payload (2xx), False otherwise.
        """
        payload: dict[str, Any] = {
            "attachments": [
                {
                    "color": "#CC0000",
                    "fallback": f"Tombstone Auto-Rollback: {flag_key} disabled in {environment}",
                    "title": "Tombstone Auto-Rollback",
                    "text": (
                        f"Flag *{flag_key}* has been automatically disabled in "
                        f"*{environment}*.\n"
                        f"Error rate: *{error_rate:.1f}%*\n"
                        f"Triggered by: {triggered_by}"
                    ),
                    "mrkdwn_in": ["text"],
                    "actions": [
                        {
                            "type": "button",
                            "text": "View in Dashboard",
                            "url": rollback_url,
                            "style": "danger",
                        }
                    ],
                }
            ]
        }
        return await self._post(payload)

    async def notify_stale_flag(
        self,
        flag_key: str,
        owner_id: str,
        days_at_100_pct: int,
        recommended_action: str,
    ) -> bool:
        """Post an amber/red attachment for a stale flag requiring cleanup.

        Args:
            flag_key: The stale flag key.
            owner_id: User or team responsible for this flag.
            days_at_100_pct: How many consecutive days the flag has been at 100%.
            recommended_action: Human-readable suggested action (e.g. "Archive flag").

        Returns:
            True if Slack accepted the payload, False otherwise.
        """
        color = "#FF4500" if days_at_100_pct >= 30 else "#FFA500"
        payload: dict[str, Any] = {
            "attachments": [
                {
                    "color": color,
                    "fallback": f"Tombstone Stale Flag: {flag_key} ({days_at_100_pct} days at 100%)",
                    "title": "Tombstone Stale Flag Detected",
                    "fields": [
                        {"title": "Flag Key", "value": flag_key, "short": True},
                        {"title": "Owner", "value": owner_id, "short": True},
                        {
                            "title": "Days at 100%",
                            "value": str(days_at_100_pct),
                            "short": True,
                        },
                        {
                            "title": "Recommended Action",
                            "value": recommended_action,
                            "short": False,
                        },
                    ],
                    "footer": "Tombstone Intelligence",
                    "mrkdwn_in": ["fields"],
                }
            ]
        }
        return await self._post(payload)

    async def notify_incident_correlation(
        self,
        incident_id: str,
        flag_key: str,
        correlation_score: float,
        rollback_url: str,
    ) -> bool:
        """Post a purple attachment when a flag correlates with an active incident.

        Args:
            incident_id: PagerDuty / OpsGenie incident identifier.
            flag_key: Flag that may be causally related to the incident.
            correlation_score: Confidence score (0.0 – 1.0).
            rollback_url: One-click rollback URL in the Tombstone dashboard.

        Returns:
            True if Slack accepted the payload, False otherwise.
        """
        payload: dict[str, Any] = {
            "attachments": [
                {
                    "color": "#6A0DAD",
                    "fallback": (
                        f"Tombstone Incident Correlation: incident {incident_id} "
                        f"may be related to flag {flag_key}"
                    ),
                    "title": "Tombstone Incident Correlation Detected",
                    "text": (
                        f"Incident *{incident_id}* correlates with flag *{flag_key}*.\n"
                        f"Correlation score: *{correlation_score:.2f}*"
                    ),
                    "mrkdwn_in": ["text"],
                    "actions": [
                        {
                            "type": "button",
                            "text": "Rollback Flag",
                            "url": rollback_url,
                            "style": "danger",
                        }
                    ],
                    "footer": "Tombstone Intelligence",
                }
            ]
        }
        return await self._post(payload)

    async def _post(self, payload: dict[str, Any]) -> bool:
        """HTTP POST the payload to the configured Slack Incoming Webhook URL.

        Args:
            payload: Slack message payload dict.

        Returns:
            True on 2xx response, False on any error or non-2xx status.
        """
        if not self._webhook_url:
            logger.warning("SLACK_WEBHOOK_URL not configured; skipping notification")
            return False

        try:
            async with httpx.AsyncClient(timeout=10.0) as client:
                response = await client.post(self._webhook_url, json=payload)
                response.raise_for_status()
                return True
        except httpx.HTTPStatusError as exc:
            logger.error(
                "Slack webhook returned %s: %s",
                exc.response.status_code,
                exc.response.text,
            )
            return False
        except httpx.RequestError as exc:
            logger.error("Failed to reach Slack webhook: %s", exc)
            return False

File: services/evaluator/internal/notify/slack.go
// Package notify posts best-effort Slack notifications for evaluator
// events. EVAL-2: the circuit breaker's auto-rollback path (rollback.go)
// has always disabled the flag and published a Redis event, but never
// notified anyone -- a human on-call learns a flag was auto-rolled-back
// only by noticing the dashboard or the audit log, not proactively.
//
// This mirrors services/intelligence/app/integrations/slack.py's
// SlackNotifier.notify_rollback message shape exactly (same color, title,
// text layout, and "View in Dashboard" button) -- that Python notifier is
// real, working code, just never instantiated by anything in the repo
// (confirmed via grep: zero imports of SlackNotifier anywhere). Ported to
// Go here, rather than adding a cross-service HTTP hop through
// marketplace's TriggerEvent/Dispatcher (which remain unwired and out of
// scope for this slice), because the evaluator is where the circuit
// breaker's own trip event already fires.
package notify

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"time"

	"go.uber.org/zap"
)

// isAbsoluteURL reports whether s is a well-formed, scheme-and-host URL
// (e.g. "https://dashboard.example.com/flags/x"), as opposed to a bare
// relative path (e.g. "/flags/x") or an empty string. Only http/https are
// treated as usable -- Slack renders a button URL as a plain hyperlink,
// so any other scheme (or none) has no defined behavior in a Slack client.
func isAbsoluteURL(s string) bool {
	u, err := url.Parse(s)
	if err != nil || !u.IsAbs() {
		return false
	}
	return u.Scheme == "http" || u.Scheme == "https"
}

// SlackNotifier posts to a Slack Incoming Webhook. Best-effort only: a
// failed or unconfigured webhook logs a warning and returns, it never
// blocks or retries the caller -- matches this codebase's existing
// fail-open telemetry/Rekor-submission philosophy for non-critical,
// third-party notification paths.
type SlackNotifier struct {
	webhookURL string
	httpClient *http.Client
	logger     *zap.Logger
}

// NewSlackNotifier constructs a notifier for the given webhook URL. An
// empty webhookURL is valid -- every notify call becomes a logged no-op,
// matching the Python SlackNotifier's own "not configured" behavior
// rather than requiring every deployment to set SLACK_WEBHOOK_URL.
func NewSlackNotifier(webhookURL string, logger *zap.Logger) *SlackNotifier {
	return &SlackNotifier{
		webhookURL: webhookURL,
		httpClient: &http.Client{Timeout: 10 * time.Second},
		logger:     logger,
	}
}

// slackAttachment/slackAction/slackPayload mirror exactly the JSON shape
// services/intelligence/app/integrations/slack.py's notify_rollback builds
// -- same field names, same nesting, so this is a faithful Go port, not a
// reinterpretation.
type slackAction struct {
	Type  string `json:"type"`
	Text  string `json:"text"`
	URL   string `json:"url"`
	Style string `json:"style"`
}

type slackAttachment struct {
	Color    string        `json:"color"`
	Fallback string        `json:"fallback"`
	Title    string        `json:"title"`
	Text     string        `json:"text"`
	MrkdwnIn []string      `json:"mrkdwn_in"`
	Actions  []slackAction `json:"actions,omitempty"`
}

type slackPayload struct {
	Attachments []slackAttachment `json:"attachments"`
}

// NotifyRollback posts a red attachment announcing an auto-rollback event.
//
// errorRate is a FRACTION in [0,1] (matching circuit.Breaker.ErrorRate's
// own return convention), converted to a 0-100 percentage for display --
// NOT the same scale as the Python notify_rollback's own docstring, which
// documents its error_rate parameter as already 0-100. Passing this
// function's errorRate straight through unconverted would silently
// display "0.1%" for what is actually a 10% error rate.
func (n *SlackNotifier) NotifyRollback(
	ctx context.Context,
	flagKey, environment string,
	errorRate float64,
	triggeredBy, rollbackURL string,
) bool {
	if n.webhookURL == "" {
		n.logger.Warn("SLACK_WEBHOOK_URL not configured; skipping rollback notification",
			zap.String("flag", flagKey), zap.String("env", environment))
		return false
	}

	attachment := slackAttachment{
		Color: "#CC0000",
		Fallback: fmt.Sprintf(
			"Tombstone Auto-Rollback: %s disabled in %s", flagKey, environment,
		),
		Title: "Tombstone Auto-Rollback",
		Text: fmt.Sprintf(
			"Flag *%s* has been automatically disabled in *%s*.\n"+
				"Error rate: *%.1f%%*\nTriggered by: %s",
			flagKey, environment, errorRate*100, triggeredBy,
		),
		MrkdwnIn: []string{"text"},
	}
	// Slack's button "url" has no documented support for relative paths --
	// a client has no base to resolve one against, so a non-absolute
	// rollbackURL (e.g. DASHBOARD_URL unset upstream, producing a bare
	// "/flags/{key}") would ship a non-functional "View in Dashboard"
	// button while NotifyRollback still reports success (Slack's webhook
	// endpoint accepts any string in "url", it doesn't validate it's a
	// working link). Omitting the button entirely when there's no usable
	// destination is honest; a broken CTA on the alert whose whole point
	// is a one-click path to the incident is not.
	if isAbsoluteURL(rollbackURL) {
		attachment.Actions = []slackAction{
			{Type: "button", Text: "View in Dashboard", URL: rollbackURL, Style: "danger"},
		}
	}

	payload := slackPayload{Attachments: []slackAttachment{attachment}}

	body, err := json.Marshal(payload)
	if err != nil {
		n.logger.Warn("marshal slack rollback payload failed", zap.Error(err))
		return false
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, n.webhookURL, bytes.NewReader(body))
	if err != nil {
		n.logger.Warn("build slack webhook request failed", zap.Error(err))
		return false
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := n.httpClient.Do(req)
	if err != nil {
		n.logger.Warn("slack webhook request failed", zap.Error(err),
			zap.String("flag", flagKey), zap.String("env", environment))
		return false
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		n.logger.Warn("slack webhook returned an error status",
			zap.Int("status", resp.StatusCode),
			zap.String("flag", flagKey), zap.String("env", environment))
		return false
	}
	return true
}
